### PR TITLE
Improve sublime.py type hints

### PR DESCRIPTION
**Pull request by AmjadHD**

### CODE REVIEW

File: sublime.pyi [hunk 1]
```diff
@@ -0,0 +1,713 @@
...
+
+    def cover(self, rhs: 'Region') -> 'Region': ...
+
+    def intersection(self, rhs: 'Region') -> 'Region': ...
+
+    def intersects(self, rhs: 'Region') -> bool: ...
+
+
+class Selection(Sized):
+    view_id: Any
```
Comment (jfcherng) [inline sublime.pyi]: `int`?

File: sublime.pyi [hunk 2]
```diff
@@ -0,0 +1,713 @@
...
+    def set_encoding(self, encoding_name: str) -> None: ...
+
+    def line_endings(self) -> str: ...
+
+    def set_line_endings(self, line_ending_name: str) -> None: ...
+
+    def size(self) -> int: ...
+
+    def begin_edit(self, edit_token, cmd,
+                   args: Optional[Dict[str, Any]] = ...): ...
```
Comment (jfcherng) [inline sublime.pyi]: `edit_token: int`
`cmd: str`

https://forum.sublimetext.com/t/new-begin-edit-signature-in-st3/8611/10?u=jfcherng

---

return `Edit` according to the implementation in `sublime.py`

File: sublime.pyi [hunk 3]
```diff
@@ -0,0 +1,713 @@
...
+
+    def window_tolayout(self, xy: _Vector) -> _Vector: ...
+
+    def window_to_text(self, xy: _Vector) -> int: ...
+
+    def line_height(self) -> float: ...
+
+    def em_width(self) -> float: ...
+
+    def is_folded(self, sr) -> bool: ...
```
Comment (jfcherng) [inline sublime.pyi]: sr: `Region` according to console error log

```
>>> view.is_folded(5)
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "C:\Program Files\Sublime Text 3\sublime.py", line 1041, in is_folded
    return sublime_api.view_is_folded(self.view_id, sr)
TypeError: Region required
```

File: sublime.pyi [hunk 4]
```diff
@@ -0,0 +1,713 @@
...
+
+    def viewport_extent(self) -> _Vector: ...
+
+    def layout_extent(self) -> _Vector: ...
+
+    def text_tolayout(self, tp: int) -> _Vector: ...
+
+    def layout_to_text(self, xy: _Vector): ...
+
+    def window_tolayout(self, xy: _Vector) -> _Vector: ...
```
Comment (jfcherng) [inline sublime.pyi]: window_to_layout

File: sublime.pyi [hunk 5]
```diff
@@ -0,0 +1,713 @@
...
+    def viewport_position(self) -> _Vector: ...
+
+    def set_viewport_position(
+        self, xy: _Vector, animate: bool = ...) -> None: ...
+
+    def viewport_extent(self) -> _Vector: ...
+
+    def layout_extent(self) -> _Vector: ...
+
+    def text_tolayout(self, tp: int) -> _Vector: ...
```
Comment (jfcherng) [inline sublime.pyi]: text_to_layout

File: sublime.pyi [hunk 6]
```diff
@@ -0,0 +1,713 @@
...
+    def rowcol(self, tp: int) -> Tuple[int, int]: ...
+
+    def text_point(self, row: int, col: int) -> int: ...
+
+    def visible_region(self) -> Region: ...
+
+    def show(self, x: Union[Selection, Region, int],
+             show_surrounds: bool = ...) -> None: ...
+
+    def show_at_center(self, x: Union[Selection, Region, int]) -> None: ...
```
Comment (jfcherng) [inline sublime.pyi]: `Selection` may not be accepted.
https://github.com/AmjadHD/sublime_dev/pull/1#discussion_r348435210

File: sublime.pyi [hunk 7]
```diff
@@ -0,0 +1,713 @@
...
+
+    def add_regions(self, key: str, regions: List[Region], scope: str = ...,
+                    icon: str = ..., flags: int = ...) -> None: ...
+
+    def get_regions(self, key: str) -> List[Region]: ...
+
+    def erase_regions(self, key: str) -> None: ...
+
+    def add_phantom(self, key: str, region: Region, content: str,
+                    layout, on_navigate: Optional[Any] = ...): ...
```
Comment (jfcherng) [inline sublime.pyi]: on_navigate: `Optional[Callable[[str], None]]`?

File: sublime.pyi [hunk 8]
```diff
@@ -0,0 +1,713 @@
...
+class Region:
+    a: int
+    b: int
+    xpos: int
+
+    def __init__(self, a: int, b: Optional[int] = ..., xpos: int = ...) -> None: ...
+
+    def __len__(self) -> int: ...
+
+    def __eq__(self, rhs) -> bool: ...
```
Comment (jfcherng) [inline sublime.pyi]: rhs: `'Region'` as well?

File: sublime.pyi [hunk 9]
```diff
@@ -0,0 +1,713 @@
...
+
+    def contains(self, region: Region) -> None: ...
+
+
+class Sheet:
+    sheet_id: int
+
+    def __init__(self, id: int) -> None: ...
+
+    def __eq__(self, other) -> bool: ...
```
Comment (jfcherng) [inline sublime.pyi]: rhs: `'Sheet'`?

File: sublime.pyi [hunk 10]
```diff
@@ -0,0 +1,713 @@
...
+class View:
+    view_id: int
+    selection: Selection
+    settings_object: 'Settings'
+
+    def __init__(self, id: int) -> None: ...
+
+    def __len__(self) -> int: ...
+
+    def __eq__(self, other) -> bool: ...
```
Comment (jfcherng) [inline sublime.pyi]: rhs: `'View'`?

File: sublime.pyi (hunk 8 above)
Comment (AmjadHD) [reply sublime.pyi]: `__eq__` should handle arbitrary objects https://stackoverflow.com/questions/37557411/why-does-defining-the-argument-types-for-eq-throw-a-mypy-type-error
I'm pretty sure i came across this before but i can't find it.

File: sublime.pyi (hunk 9 above)
Comment (AmjadHD) [reply sublime.pyi]: `__eq__` should not be annotated.

File: sublime.pyi (hunk 10 above)
Comment (AmjadHD) [reply sublime.pyi]: `__eq__` should not be annotated

File: sublime.pyi (hunk 6 above)
Comment (AmjadHD) [reply sublime.pyi]: Yeah i copied the signature from `show`, by the way, should it accept a `Selection` ?

Comment (jfcherng) [reply sublime.pyi]: for `show()`, yes. `view.show(view.sel())` executes without an error/warning

Comment (AmjadHD) [reply sublime.pyi]: Yes, but does it make sense ?

Comment (jfcherng) [reply sublime.pyi]: what do you mean? you cannot use `Selection` in `show_at_center` anyway at this moment

Comment (AmjadHD) [reply sublime.pyi]: I mean in `show`, are there any real world usages of `show` with `Selection` ?

Comment (jfcherng) [reply sublime.pyi]: idk. it's counter-intuitive but yes people can do that.

**Comment by FichteFoll** [comment]
Maybe adopt https://github.com/SublimeText/sublime_lib/blob/master/stubs/sublime.pyi ? I haven't checked which is more complete or correct.

File: sublime.pyi [hunk 11]
```diff
@@ -0,0 +1,713 @@
...
+    def set_project_data(self, v: Dict[str, _Value]) -> None: ...
+
+    def settings(self) -> 'Settings': ...
+
+    def template_settings(self) -> 'Settings': ...
+
+    def lookup_symbol_in_index(self, sym: str) -> List[_Location]: ...
+
+    def lookup_symbol_in_open_files(self, sym: str) -> List[_Location]: ...
+
```
Comment (jfcherng) [inline sublime.pyi]: 2 more methods.

```py
    def lookup_references_in_index(self, sym: str) -> List[_Location]: ...
    def lookup_references_in_open_files(self, sym: str) -> List[_Location]: ...
```

File: sublime.pyi [hunk 12]
```diff
@@ -0,0 +1,713 @@
...
+    def layout_extent(self) -> _Vector: ...
+
+    def text_tolayout(self, tp: int) -> _Vector: ...
+
+    def layout_to_text(self, xy: _Vector): ...
+
+    def window_tolayout(self, xy: _Vector) -> _Vector: ...
+
+    def window_to_text(self, xy: _Vector) -> int: ...
+
```
Comment (jfcherng) [inline sublime.pyi]: 2 more methods.

```py
    def text_to_window(self, tp: int) -> _Vector: ...
    def layout_to_window(self, xy: _Vector) -> _Vector: ...
```

**Comment by jfcherng** [comment]
Yeah. After diff with the one from `sublime_lib`, I feel `sublime_lib`'s is generally better at this moment.

File: sublime.pyi (hunk 2 above)
Comment (AmjadHD) [reply sublime.pyi]: done!

File: sublime.pyi (hunk 1 above)
Comment (AmjadHD) [reply sublime.pyi]: done!

File: sublime.pyi (hunk 3 above)
Comment (AmjadHD) [reply sublime.pyi]: done!

File: sublime.pyi (hunk 4 above)
Comment (AmjadHD) [reply sublime.pyi]: fixed!

File: sublime.pyi (hunk 5 above)
Comment (AmjadHD) [reply sublime.pyi]: fixed!

File: sublime.pyi (hunk 6 above)
Comment (AmjadHD) [reply sublime.pyi]: fixed!

File: sublime.pyi (hunk 7 above)
Comment (AmjadHD) [reply sublime.pyi]: done!

**Comment by AmjadHD** [comment]
Actually I did a diff and found that sublime_dev is more complete (not by far though), it's type hints are more precise, but I liked the idea of `callback0` and `callback1` maybe use them ?
@jfcherng, @FichteFoll, Do you think `_Location`, `_Vector`, `_Layout` ... should be made public (i.e remove underscores) ?

File: sublime.pyi (hunk 11 above)
Comment (AmjadHD) [reply sublime.pyi]: done!

File: sublime.pyi (hunk 12 above)
Comment (AmjadHD) [reply sublime.pyi]: done!

**Comment by FichteFoll** [comment]
I don't know about any conventions for whether or not type aliases should be public or not, but it seems type variables (`_T`) are generally private.

**Comment by FichteFoll** [comment]
You should run mypy with the following config to spot some errors or missing type hints:

```
[mypy]
ignore_missing_imports = True
strict_optional = True
# warn_unused_ignores = True
warn_incomplete_stub = True
warn_redundant_casts = True

check_untyped_defs = True
disallow_subclassing_any = True
disallow_untyped_defs = True
# warn_return_any = True
```

**Comment by AmjadHD** [comment]
@FichteFoll, can you give me the command ?

**Pull request merged**